tests/qxmppatmmanager/tst_qxmppatmmanager.cpp

```diff
@@ -0,0 +1,1258 @@
+#include <QObject>
+#include <QSet>
+
+class tst_QXmppAtmManager : public QObject
+{
+    Q_OBJECT
+
+signals:
+    void resultProcessed();
+    void unexpectedTrustMessageSent();
+
+private slots:
+    void initTestCase();
+    void testSendTrustMessage();
+    void testMakeTrustDecisionsOwnKeys();
+    void testMakeTrustDecisionsOwnKeysNoOwnEndpoints();
+    void testMakeTrustDecisionsOwnKeysNoOwnEndpointsWithAuthenticatedKeys();
+    void testMakeTrustDecisionsOwnKeysNoContactsWithAuthenticatedKeys();
+    void testMakeTrustDecisionsSoleOwnKeyDistrusted();
+    void testMakeTrustDecisionsContactKeys();
+    void testMakeTrustDecisionsContactKeysNoOwnEndpoints();
+    void testMakeTrustDecisionsContactKeysNoOwnEndpointsWithAuthenticatedKeys();
+    void testMakeTrustDecisionsSoleContactKeyDistrusted();
+    void testHandleMessageReceived();
+
+private:
+    void testMakeTrustDecisionsOwnKeysDone();
+    void testMakeTrustDecisionsContactKeysDone();
+
+    QXmppClient m_client;
+    QXmppLogger m_logger;
+    QXmppAtmManager *m_manager;
+    QXmppTrustMemoryStorage *m_trustStorage;
+    QXmppCarbonManager *m_carbonManager;
+};
+
+void tst_QXmppAtmManager::initTestCase()
+{
+    m_trustStorage = new QXmppTrustMemoryStorage;
+    m_manager = new QXmppAtmManager(m_trustStorage);
+    m_client.addExtension(m_manager);
+    m_client.configuration().setJid("alice@example.org");
+
+    m_carbonManager = new QXmppCarbonManager;
+    m_carbonManager->setCarbonsEnabled(true);
+    m_client.addExtension(m_carbonManager);
+
+    m_logger.setLoggingType(QXmppLogger::SignalLogging);
+    m_client.setLogger(&m_logger);
+}
+
+void tst_QXmppAtmManager::testSendTrustMessage()
+{
+    QXmppTrustMessageKeyOwner keyOwnerAlice;
+    keyOwnerAlice.setJid(QStringLiteral("alice@example.org"));
+    keyOwnerAlice.setTrustedKeys({ QStringLiteral("d11715b069372e7a4416caaaced4f30200838155e57dad37a5a4aa24538e58e5"),
+                                   QStringLiteral("b589ffc1c20ec414a85b85b551f3ebff381b2e2a412b62ac15f0bb1756f3badd") });
+    keyOwnerAlice.setDistrustedKeys({ QStringLiteral("788a40d0eae5a40409d4687a36d3106bbe361971aec0245598571e7b629edc6b"),
+                                      QStringLiteral("b6c21e111bd4f9ed06ee0485cb302bf1238e90b89986a04061e48d49ddbe95cb") });
+
+    QXmppTrustMessageKeyOwner keyOwnerBob;
+    keyOwnerBob.setJid(QStringLiteral("bob@example.com"));
+    keyOwnerBob.setTrustedKeys({ QStringLiteral("d11715b069372e7a4416caaaced4f30200838155e57dad37a5a4aa24538e58e5"),
+                                 QStringLiteral("b589ffc1c20ec414a85b85b551f3ebff381b2e2a412b62ac15f0bb1756f3badd") });
+    keyOwnerBob.setDistrustedKeys({ QStringLiteral("788a40d0eae5a40409d4687a36d3106bbe361971aec0245598571e7b629edc6b"),
+                                    QStringLiteral("b6c21e111bd4f9ed06ee0485cb302bf1238e90b89986a04061e48d49ddbe95cb") });
+
+    const auto keyOwners = { keyOwnerAlice, keyOwnerBob };
+
+    const auto *context = new QObject(this);
```

Comment:
Just stack-allocate `context`, so you can avoid the manual deletion of it.
```suggestion
    QObject context;
```